man/fstab-decode.8
'\" -*- coding: UTF-8 -*-
.\" A man page for fstab-decode(8).
.\"
.\" Copyright (C) 2006 Red Hat, Inc. All rights reserved.
.\"
.\" This copyrighted material is made available to anyone wishing to use,
.\" modify, copy, or redistribute it subject to the terms and conditions of the
.\" GNU General Public License v.2.
.\"
.\" This program is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
.\" more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
.\"
.\" Author: Miloslav Trmac <mitr@redhat.com>
.TH FSTAB-DECODE 8 "May 2006" "sysvinit " "Linux System Administrator's Manual"

.SH NAME
fstab-decode \- run a command with fstab-encoded arguments

.SH SYNOPSIS
.B fstab-decode
.IR COMMAND " [" ARGUMENT " \&.\|.\|.\&]"

.SH DESCRIPTION
.B fstab-decode
decodes escapes (such as newline characters and other whitespace)
in the specified
.IR ARGUMENT s
and uses them to run
.IR COMMAND .
The argument escaping uses the same rules as path escaping in
.IR /etc/fstab ", " /etc/mtab " and " /proc/mtab .

In essence \fBfstab-decode\fR can be used anytime we want to pass multiple
parameters to a command as a list of command line arguments.
It turns output like this:

.nf
/root
/mnt/remote-disk
/home
.fi

Into one long list of parameters, "/root /mnt/remote-disk /home".
This can be useful when trying to work with multiple filesystems at once.
For instance, we can use it to unmount multiple NFS shares.
This program also removes whitespace and
other characters which might cause programs such as
\fBmount\fR(8) or \fBumount\fR(8) to fail.

.SH EXIT STATUS
.B fstab-decode
exits with status 127 if
.I COMMAND
can't be run.
Otherwise it exits with the status returned by
.IR COMMAND .

.SH EXAMPLES
The following example reads
.IR fstab ,
finds all instances of VFAT filesystems
and prints their mount points
(argument 2 in the
.I fstab
file).
.B fstab-decode
then runs the specified program, \fBumount\fR(8), and passes
it the list of VFAT mountpoints.
This unmounts all VFAT partitions.

.nf
.B fstab-decode umount $(awk \[aq]$3 == \[dq]vfat\[dq] { print $2 }\[aq] \
/etc/fstab)
.fi

.SH SEE ALSO
.BR fstab (5)


doc/Changelog
sysvinit (3.11) unreleased; urgency=low

    * Some escape characters were included in the inittab manual page, but not displayed
    by the "man" command because they were not (ironically) properly escaped. This has been
    fixed.

    * Enabled chaining commands together in the inittab file. This allows the admin
      to run commands like "task1 && task2" or "task2 || task2" from the inittab file.

    * Fix typoes in halt manual page. Fixes provided by Bjarni Ingi Gislason.

    * Fix typos/markdown in fstab-decode manual page.
      Patch provided by Bjarni Ingi Gislason.


sysvinit (3.10) released; urgency=low

    *  When the user executes "machinectl stop", systemd sends SIGRTMIN+4 to PID 1 
       in the container, and expects that to initiate a graceful shutdown (power-off).
       SysV init now catches this signal and initiates a shutdown (shutdown -hP now).
       - floppym provided patch to accomplish this.

    * Fix issue in bootlogd which could cause the service to enter an endless loop
      (and use too much CPU) when it is able to open a device for writing, but not actually
      able to write to it. This resulted in bootlogd closing and re-opening the device over
      and over. Now bootlogd should simply fail gracefully when it cannot write to an open
      file/device.

    * Fix formatting in shutdown.8 manual page. Cleaned up whitespace and special characters.


sysvinit (3.09) released; urgency=low

    * Patch for man/Makefile to fix the clean recipe.
      Provided by Lucas Nussabaum and Mark Hindley

    * On Linux systems, allow reboot command to pass a message
      to the system firmware during the restart. This is
      accomplished with the -m flag.

    * Patch from kraj which allows hddown to compile
      when musl is the C library.

sysvinit (3.08) released; urgency=low
    * floppym provided patch which causes the halt command
      to call "shutdown -h -H" instead of "shutdown -h" when
      halt is invoked without parameters. This forces the shutdown
      command to set the INIT_HALT variable and assume, unless other
      conditions apply, that the "halt" call really wants to halt the
      machine and INIT_HALT should be set. In other words we
      assume halt wants to halt unless told otherwise.
      Addresses downstream Gentoo bug ID 911257.

    * Applied a patch from floppm which adds kexec option to the halt command.

    * Updated halt documentation and help output to display parameters in
      alphabetical order.


sysvinit (3.07) released; urgency=low
    * Fixed killall5 so that processes in the omit list are
      not sent any signals, including SIGSTOP.
    * Fixed usage message for killall5 to be more accurate.
    * pidof was not returning PIDs of programs which were launched
      using a symbolic link. (ie /tmp/sleep when /tmp/sleep links to /usr/bin/sleep).
      This is now fixed as we check both the realpath and symbolic path for processes.
      In other words, "pidof /tmp/sleep" and "pidof /usr/bin/sleep" will return
      the same PIDs when /tmp/sleep is a symbolic link to /usr/bin/sleep.
    * Fixed memory initialization error in pidof. Fix provided by Markus Fischer.
    * Accepted patch from Mark Hindley  which avoids clearing realpath information
      in pidof when trying to find matching executables.


sysvinit (3.06) released; urgency=low
    * Mark Hindley fixed typo in es.po
    * Mark Hindley cleaned up translation code in src/Makefile. 
    * Drop sulogin from Debian build. Removed libcrypt-dev dependency.
    * Fixed pt translation pages which were failing due to mis-matched
      open/close tags.
    * Makefile now respects ROOT prefix when setting up pidof-to-killall5
      symbolic link.
    * Removed redundant translation files from man directory.
    * Makefile now respects DESTDIR. User can specify either ROOT= or DESTDIR=
      to set install prefix.


sysvinit (3.05) released; urgency=low
    * Helge Kreutzmann provided updated Makefile for
      translation of manual pages. This has been added
      to the man directory.
    * Added sys/sysmacros.h include in mountpoint.c to fix
      compiler errors on systems where major/minor macros were not defined.
    * Applied patches from Mark Hindley to clean up man page
      Makefile, translations and installs of new man pages.
    * Remove reliance on linux/fs.h as it conflicts with glibc 2.36.
      Patch provided by lucascars.

sysvinit (3.04) released; urgency=low
    * Mark Hindley supplied patch to make bootlogd compile
      on GNU Hurd systems. Was missing major/minor macro.

sysvinit (3.03) released; urgency=low
    * Fixed formatting in init.8 man page.
      Patch provided by Mark Hindley.
    * Fixed compile error on GNU Hurd and musl-base systems. Due to missing MAX_PATH definition.
      Error reported by Mark Hindly.

sysvinit (3.02) released; urgency=low
    * Added q and Q flags to synopsis in shutdown manual page.
    * Applied fixes for markup and spacing in manual pages.
      Patch provided by Mario Blattermann.
    * Added translation framework (po4a) from Mario Blttermann.
    * Added Makefile for man/ directory. Will handle translations
      and substitutions.
    * Applied new translations for multiple languages from Mario Blattermann.
    * Added ability to use "@" symbol in command named in the inittab file. This
      treats commands as literal and does not launch a shell to interpret them.
    * Updated inittab manual page to include overview of symbols which trigger
      a shell interpretor and how to disable them using the @ symbol.
    * Introduced change which adds error checking in bootlogd when performing
      chdir(). - Provided by Alexander Vickberg 
    * Add check for console using TIOCGDEV on Linux systems in bootlogd to
      make finding console more robust. - Provided by Alexander Vickberg
 

sysvinit (3.01) released; urgency=low
    * Default to showing processes in the uninterruptable state (D).

src/fstab-decode.c
/* fstab-decode(8).

Copyright (c) 2006 Red Hat, Inc. All rights reserved.

This copyrighted material is made available to anyone wishing to use, modify,
copy, or redistribute it subject to the terms and conditions of the GNU General
Public License v.2.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
Street, Fifth Floor, Boston, MA 02110-1301, USA.

Author: Miloslav Trmac <mitr@redhat.com> */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* Decode the fstab-encoded string in place. */
static void
decode(char *s)
{
	const char *src;
	char *dest;

	src = s;
	dest = s;
	while (*src != '\0') {
		if (*src != '\\')
			*dest = *src++;
		else {
			static const struct repl {
				char orig[4];
				size_t len;
				char new;
			} repls[] = {
#define R(X, Y) { X, sizeof(X) - 1, Y }
				R("\\", '\\'),
				R("011", '\t'),
				R("012", '\n'),
				R("040", ' '),
				R("134", '\\')
#undef R
			};

			size_t i;

			for (i = 0; i < sizeof (repls) / sizeof (repls[0]);
			     i++) {
				if (memcmp(src + 1, repls[i].orig,
					   repls[i].len) == 0) {
					*dest = repls[i].new;
					src += 1 + repls[i].len;
					goto found;
				}
			}
			*dest = *src++;
		found:
			;
		}
		dest++;
	}
	*dest = '\0';
}

int
main (int argc, char *argv[])
{
	size_t i;

	if (argc < 2) {
		fprintf(stderr, "Usage: fstab-decode command [arguments]\n");
		return EXIT_FAILURE;
	}
	for (i = 2; i < (size_t)argc; i++)
		decode(argv[i]);
	execvp(argv[1], argv + 1);
	fprintf(stderr, "fstab-decode: %s: %s\n", argv[1], strerror(errno));
	return 127;
}
